airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,482 @@
+    ):
+        """Constructor
+
+        :param method:
+        :param url:
+        :param params:
+        :param headers:
+        """
+        self._method = method
+        self._url = url
+        self._params = {str(k): str(v) for k, v in (params or {}).items()}
+        self._headers = {str(k): str(v) for k, v in (headers or {}).items()}
+
+    @staticmethod
+    def _match_dict(obj: Mapping[str, Any], pattern: Mapping[str, Any]) -> bool:
+        """Check that all elements from pattern dict present and have the same values in obj dict
+
+        :param obj:
+        :param pattern:
+        :return:
+        """
+        return pattern.items() <= obj.items()
+
+    def __call__(self, request: Any) -> bool:
+        """
+
+        :param request:
+        :return: True if pattern matches the provided request object, False - otherwise
+        """
+        if isinstance(request, requests.Request):
+            prepared_request = request.prepare()
+        elif isinstance(request, requests.PreparedRequest):
+            prepared_request = request
+        else:
+            return False
+
+        if self._method is not None:
+            if prepared_request.method != self._method:
+                return False
+        if self._url is not None and prepared_request.url is not None:
+            url_without_params = prepared_request.url.split("?")[0]
+            if url_without_params != self._url:
+                return False
+        if self._params is not None:
+            parsed_url = parse.urlsplit(prepared_request.url)
+            params = dict(parse.parse_qsl(str(parsed_url.query)))
+            if not self._match_dict(params, self._params):
+                return False
+        if self._headers is not None:
+            if not self._match_dict(prepared_request.headers, self._headers):
+                return False
+        return True
+
+
+class BaseCallRatePolicy(AbstractCallRatePolicy, abc.ABC):
+    def __init__(self, matchers: list[RequestMatcher]):
+        self._matchers = matchers
+
+    def matches(self, request: Any) -> bool:
+        """Tell if this policy matches specific request and should apply to it
+
+        :param request:
+        :return: True if policy should apply to this request, False - otherwise
+        """
+
+        if not self._matchers:
+            return True
+        return any(matcher(request) for matcher in self._matchers)
+
+
+class FixedWindowCallRatePolicy(BaseCallRatePolicy):
+    def __init__(self, window: TimeWindow, call_limit: int, **kwargs: Any):
+        """A policy that allows {call_limit} calls within a {window} time interval
+
+        :param window:
+        :param call_limit:
+        """
+
+        self._current_window = TimeWindow(window.start, window.end)
+        self._offset = window.end - window.start
+        self._call_limit = call_limit
+        self._calls_num = 0
+        super().__init__(**kwargs)
+
+    def try_acquire(self, request: Any, weight: int) -> None:
```

Comment:
doesn't this method need to use a lock to prevent multiple thread from updating the current window and the calls_num at the same time?

Reply:
absolutely, especially for lines like 
```python
self._calls_num += 1
```
added RLock here for `try_acquire` and `update` methods